### ISSUE TITLE
Cause a browser cache reload when a setting is changed on Popup Blocker settings page

### ISSUE THREAD

**Issue by ameshkov** [AdguardTeam/PopupBlocker#131]
The original issue is described here:
https://github.com/AdguardTeam/AdguardForWindows/issues/2258

While it needs to be solved on the AG for Win side, on the PB side we should also ensure that the browser is flushing the cache.

```
/**
 * Force flush the current page cache
 */
let flushPageCache = function() {
    let xhr = new XMLHttpRequest();
    xhr.open('GET', window.location.href, true);

    xhr.setRequestHeader('Pragma', 'no-cache');
    xhr.setRequestHeader('Expires', -1);
    xhr.setRequestHeader('Cache-Control', 'no-cache');

    xhr.send();
};
```

Further testing shows that we need to bypass cache not just on the `popupblocker.adguard.com`, but on third-party websites as well.

**Third-party websites issue:** 
In addition to the first comment, we also need to force cache bypass on third-party websites.

**Steps to reproduce the issue**

1. Install PB dev build into AG
2. Open http://testcases.adguard.com/PopupBlocker/test-popup-blocker.html
3. Click "OnClick Open" and click somewhere on the page
4. Choose "Allow pop-ups on testcases.adguard.com"
5. Reload the page
6. Go to point 3

**Expected result**

The popup is opened;

**Actual result**

The popup is suppressed as if nothing happened;

Using `flushPageCache` will solve it.

Again, this is a serious issue with the userscripts implementation on AG4Win side and it must be addressed in `AdguardTeam/AdguardForWindows#2258`, but for the sake of older AG versions, we need to solve it on the PB side as well.

**Comment by seanl-adg** [comment]
I wonder if we can resolve the issue by using [Service Worker API](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API)

**Comment by ameshkov** [comment]
Another option is to add these to the `options.html` code:
```
<meta http-equiv="cache-control" content="no-cache" />
<meta http-equiv="pragma" content="no-cache" />
```

This one needs to be tested (flushing via xhr works for sure).

**Comment by ameshkov** [comment]
> I wonder if we can resolve the issue by using Service Worker API

For this particular case (PB options page) that would be an overkill.

For a generic solution for all userscripts -- maybe, any specific ideas on how it can be used?

**Comment by seanl-adg** [comment]
I wonder if smth like this would work:

```js
addEventListener('fetch', event => {
    let requestUrl = new URL(event.request.url);
    if (requestUrl.hostname.endsWith('.adguard.com')) {
        event.respondWith(async function () {
            return fetch(event.request, { cache: "no-cache" });
        });
    }
});
```

**Comment by ameshkov** [comment]
In addition to the first comment, we also need to force cache bypass on third-party websites.

## Steps to reproduce the issue

1. Install PB dev build into AG
2. Open http://testcases.adguard.com/PopupBlocker/test-popup-blocker.html
3. Click "OnClick Open" and click somewhere on the page
4. Choose "Allow pop-ups on testcases.adguard.com"
5. Reload the page
6. Go to point 3

## Expected result
The popup is opened;

## Actual result
The popup is suppressed as if nothing happened;

Using `flushPageCache` will solve it.

Again, this is a serious issue with the userscripts implementation on AG4Win side and it must be addressed in `AdguardTeam/AdguardForWindows#2258`, but for the sake of older AG versions, we need to solve it on our side.

**Comment by ameshkov** [comment]
> I wonder if smth like this would work:

I didn't test it but it looks as it should work. 

It's not too generic though. I mean if we make our AG4Win userscripts implementation use SW API to bypass the browser's cache, we'll need smth more complicated. Also, as I recall, service workers must be same-origin so we can't use our fake `local.adguard.com` domain to load its code -> we'll need to emulate a same-origin file.

**Comment by seanl-adg** [comment]
> I mean if we make our AG4Win userscripts implementation use SW API to bypass the browser's cache, we'll need smth more complicated. 

I've never said we should do that.

**Comment by ameshkov** [comment]
Mm, where's my comment about third-party websites caching?

**Comment by ameshkov** [comment]
> I've never said we should do that.

Further testing shows that we need to bypass cache not just on the `popupblocker.adguard.com`, but on third-party websites as well.

**Third-party websites issue:**
In addition to the first comment, we also need to force cache bypass on third-party websites.

## Steps to reproduce the issue

1. Install PB dev build into AG
2. Open http://testcases.adguard.com/PopupBlocker/test-popup-blocker.html
3. Click "OnClick Open" and click somewhere on the page
4. Choose "Allow pop-ups on testcases.adguard.com"
5. Reload the page
6. Go to point 3

## Expected result
The popup is opened;

## Actual result
The popup is suppressed as if nothing happened;

Using `flushPageCache` will solve it.

Again, this is a serious issue with the userscripts implementation on AG4Win side and it must be addressed in `AdguardTeam/AdguardForWindows#2258`, but for the sake of older AG versions, we need to solve it on the PB side as well.
